Rework git tutorial and slides into unified marp presentation (#65)
jmunixusers/presentations: git101/index.md

---
![height:15cm](commit-freq.png)
<https://insights.stackoverflow.com/survey/2018>

---
![height:15cm](collab-tools.png)
<https://insights.stackoverflow.com/survey/2020>

---
# So what about GitHub?
- A social network for git users
- A place to keep a backup repository copy
- A means to exchange changes between people
- A database of issues
- A database of change requests, comments, and approvals
- An API for running events on repository changes

---
# So what about GitHub?
- Remember that git is a tool, and GitHub is a provider of git services
- Other services:
  - Azure Repos, Google Cloud Source, AWS CodeCommit
  - Bitbucket
  - GitLab
- Other git implementations:
  - Eclipse EGit
  - GitHub Desktop
  - Git Bash (ok, so it's just a port)
- A [student pack](https://education.github.com/pack) is available

---
# What is git?

- Git stores a series of changes to files and attributes of that change
  - File diff, author, timestamp, previous change hash
- Change and attributes are hashed together to uniquely identify a commit
- Branches allow different authors to pursue different paths
  - Like an actual tree, branches can go their own way, but the trunk must be shared

> - Never delete a commit, use a revert to undo, but that change is then logged

Reviewer: ```suggestion
- Never delete a commit: use a revert to undo, but that change is then logged
```

- Some leeway if your changes haven't been pushed and become visible to others

---
# A note on branching and language

- Git-based projects have a primary branch, traditionally known as `master`
- You can develop new code on different branches, with the goal of integrating
that code back into the primary branch
- Terminology and language matter, and git's use of the term `master` is
considered insensitive and has been deprecated
- Consider naming your primary branch `main` or `develop` on new projects

---
# Help along the way

- Git uses man pages extensively to document its options and features
- One man page would be too much for all of git, so pages are hyphenated with
the git-command. For example:
  - `git commit` uses `git-commit(1)`
  - `git push` uses `git-push(1)`
- The `git(1)` man page exists to document the common options and provide an overview

---
# Demo Time

---
# Creating a repository on the GitHub website

Using the + button in the top right corner or the webpage, choose
"New Repository". Pick a name and optionally a description. For this tutorial,

> we will not be using a pre-provided README, .gitignore, or license.

Reviewer: ```suggestion
we will not be using a pre-provided README, .gitignore, or license.
Remember the name; you will need this later!
```

Select the HTTPS button towards the top of this guide, and notice the section

> for `push an existing repository`. Don't need to do anything with this now.

Reviewer: ```suggestion
for `push an existing repository`. No need to do anything with this now.
```

Just keep the page open.

---
# Starting your first `git` repository

First, we will create a directory to use for this tutorial.

    mkdir -p ~/UUG/uug-git-intro && cd $_

> If we try to use a git command here, we will get an error

Reviewer: ```suggestion
If we try to use a git command here, we will get an error:
```

    git commit -am "Test commit"

Notice that you got an error when you tried to do this! You cannot create a
commit in git without that directory being a part of a git repository.

---
# Starting your first `git` repository

> So let's initialize your git repository. This starts getting `git` set up to
> run.

Reviewer: ```suggestion
Let's initialize your git repository.
```

    git init

If you run `ls -al`, you will now see a `.git` directory there. This is where
git stores all of the data about your repository.

---
# Starting your first `git` repository

You now have to tell `git` where your code should be pushed to. We are using

> GitHub for this and we'll now we will use that command from earlier.

Reviewer: ```suggestion
GitHub for this. Run the following command, but change `YOURUSERNAME` to your
GitHub user name and change `REPONAME` to the repository name you chose earlier.
```

    git remote add origin https://github.com/YOURUSERNAME/REPONAME.git

Remember later that you have named your GitHub URL `origin`.

---
# Basic `git` configuration

By default, `git` will have your name and email configured to be your login
name@hostname. This is nice in some ways, but probably not your actual

> email address. Let's fix this.

Reviewer: ```suggestion
email address. Let's fix this. Make sure you substitute all of your own info
in the following commands:
```

    git config --global user.name "Your Name"
    git config --global user.email "your@email.address"
    git config --global color.ui true

Now `git` will always know who you are, and this information will be in all
your commits.

---
# Making your first commit

Commits are basically a record of changes to your directory. Each time you
make a change, you need to tell `git` to create a new record. You do this
with the commit command.

Before you commit, you have to make a change to the directory. Let's create

> a file

Reviewer: ```suggestion
a file:
```

>     vim README.txt

Reviewer: ```suggestion
    nano README.txt
```

Let's avoid having to extricate newbies from vim via Teams! :)

---
# Making your first commit

Now that we have created a file, we can ask `git` to tell us the status of
of our directory.

    git status

Normally `git` can tell exactly what's different in the file.

    git diff

But that only works against the last commit, so it will not show our README.

---
# Making your first commit

We should instruct `git` to start tracking changes to our README.txt file. We
also need to add it to our 'stage'. To do this, we will use the following command.

    git add README.txt

Let's tell `git` to create a record of our change

    git commit

Vim (or your default editor) will open. Type a description of your changes,
called a commit message, and save the file (in Vim, you can type `:wq` in
normal mode).

---
# See a record of commits

> `git` keeps a record of each of your changes in a log. TO see this, let's run

Reviewer: ```suggestion
`git` keeps a record of each of your changes in a log. To see this:
```

    git log

---
# Changing branch name

`git log` should show that git has automatically chosen the name `master` for your
primary branch. Let's change that now.

    git branch -M main

---
# Synchronizing GitHub

The first time you push to an upstream copy, you can use the `-u` option to tell
git your preferred default.

    git push -u origin main

This will tell git to push your current branch `main` (since you just renamed it)
to the `main` branch at the `origin` URL, and to remember that for future use. Now

> that you've set the default, you can simple use

Reviewer: ```suggestion
that you've set the default, you can simply use:
```

    git push

Go check GitHub now to confirm your changes were uploaded.

---
# Reverting a commit

As you're making changes, you very often will want to undo one of your previous
commits. `git` allows for this with its `revert` command.

Let's modify our directory and commit that change.

    git rm README.txt
    git commit

---
Later, we realize that we regret this change. We can undo this. The way that
reverting works in `git` is by taking the state of the repository at a point in
time and bringing it back to the present.

So what we need to do is check our log, and find the hash of the commit we want

> to revert

Reviewer: ```suggestion
to revert. Use the following command (replacing the hash):
```

    git log
    git revert "hash"

This will create a commit with the default message, 'Revert "Old Commit"'. Push

> this change back to GitHub with

Reviewer: ```suggestion
this change back to GitHub:
```

    git push

---

> # gitignore

Reviewer: ```suggestion
# gitignore

```

- While git can contain binary files (like the images in this presentation),
tradition says it should not contain binaries that are artifacts of source files

>   - Commit your C or Java files, not executables, .o, or .class files

Reviewer: ```suggestion
  - Commit your C or Java files, not executables, `.o`, or `.class` files
```

- Other services:
  - Azure Repos, Google Cloud Source, AWS CodeCommit
  - Bitbucket
  - GitLab
- Other git implementations:
  - Eclipse EGit
  - GitHub Desktop
  - Git Bash (ok, so it's just a port)
- A [student pack](https://education.github.com/pack) is available

---
# What is git?

- Git stores a series of changes to files and attributes of that change
  - File diff, author, timestamp, previous change hash
- Change and attributes are hashed together to uniquely identify a commit
- Branches allow different authors to pursue different paths
  - Like an actual tree, branches can go their own way, but the trunk must be shared
- Never delete a commit: use a revert to undo, but that change is then logged
- Some leeway if your changes haven't been pushed and become visible to others

---
# A note on branching and language

- Git-based projects have a primary branch, traditionally known as `master`
- You can develop new code on different branches, with the goal of integrating
that code back into the primary branch
- Terminology and language matter, and git's use of the term `master` is
considered insensitive and has been deprecated

> - Consider naming your primary branch `main` or `develop` on new projects

Reviewer 2: Is Git 2.28 in Ubuntu 20.04? https://github.blog/2020-07-27-highlights-from-git-2-28/#introducing-init-defaultbranch

Author: 2.25 https://packages.ubuntu.com/focal/git

---
# Help along the way

- Git uses man pages extensively to document its options and features

> - One man page would be too much for all of git, so pages are hyphenated with

Reviewer 2: Each command also has a `--help` flag which usually opens up that particular commands `man` page.

Author: No GNU options allowed here
